raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -565,20 +583,62 @@ private FetchQuorumRecordsResponseData buildEmptyFetchQuorumRecordsResponse(
      * - {@link Errors#INVALID_REQUEST} if the request epoch is larger than the leader's current epoch
      *     or if either the fetch offset or the last fetched epoch is invalid
      */
-    private FetchQuorumRecordsResponseData handleFetchQuorumRecordsRequest(
+    private CompletableFuture<FetchQuorumRecordsResponseData> handleFetchQuorumRecordsRequest(
         RaftRequest.Inbound requestMetadata
     ) throws IOException {
         FetchQuorumRecordsRequestData request = (FetchQuorumRecordsRequestData) requestMetadata.data;
+        FetchQuorumRecordsResponseData response = tryCompleteFetchQuorumRecordsRequest(request);
+        if (response.errorCode() != Errors.NONE.code()
+            || response.records().hasRemaining()
+            || request.maxWaitTimeMs() == 0) {
+            return completedFuture(response);
+        }
+
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        purgatory.await(future, request.maxWaitTimeMs());
+
+        return future.handle((nil, exception) -> {
+            if (exception != null) {
+                Throwable cause = exception instanceof ExecutionException ?
+                    exception.getCause() : exception;
+
+                // If the fetch timed out in purgatory, it means no new data is available,
+                // and we will complete the fetch successfully. Otherwise, if there was
+                // any other error, we need to return it.
+                Errors error = Errors.forException(cause);
+                if (error != Errors.REQUEST_TIMED_OUT) {
+                    return buildEmptyFetchQuorumRecordsResponse(error, OptionalLong.empty());
+                }
+            }
+
+            try {
+                return tryCompleteFetchQuorumRecordsRequest(request);
+            } catch (Exception e) {
+                logger.error("Caught unexpected error in fetch completion of request {}", request, e);
+                return buildEmptyFetchQuorumRecordsResponse(Errors.UNKNOWN_SERVER_ERROR, OptionalLong.empty());
+            }
+        });
+    }
+
+    private FetchQuorumRecordsResponseData tryCompleteFetchQuorumRecordsRequest(
+        FetchQuorumRecordsRequestData request
+    ) throws IOException {
         Optional<Errors> errorOpt = validateLeaderOnlyRequest(request.leaderEpoch());
         if (errorOpt.isPresent()) {
-            return buildEmptyFetchQuorumRecordsResponse(errorOpt.get(), OptionalLong.empty());
+            return buildFetchQuorumRecordsResponse(
+                errorOpt.get(),
+                MemoryRecords.EMPTY,
+                OptionalLong.empty());
         }
 
         int replicaId = request.replicaId();
         long fetchOffset = request.fetchOffset();
         int lastFetchedEpoch = request.lastFetchedEpoch();
 
-        if (fetchOffset < 0 || lastFetchedEpoch < 0 || lastFetchedEpoch > request.leaderEpoch()) {
+        if (request.maxWaitTimeMs() < 0
```

Comment:
Ditto here, this check is also static and not rely on local data, we can extract it out too.

Reply:
True, these checks only depend on the request, so we could move them. I guess the general order of validation could be the following:

1. Check request sanity
2. Check epoch and quorum state
3. Handle request

Currently I had 1 and 2 reversed, but I don't think there's a strong reason to do it one way or the other.